### PR TITLE
add is_diagonal utility and Transform property

### PR DESCRIPTION
**Pull request by grlee77**
# Description

This small PR was split off from #4303 as requested by @andy-sweet for ease of review. It adds an `is_diagonal` property to Transforms that makes `Affine` transforms more efficient in the case when they are known to be diagonal.

I chose to ignore any entries below a certain tolerance when determining if a matrix is diagonal. I think in practice it is desirable as things like applying a rotation followed by an opposite rotation on a diagonal Affine, should ideally return the original Affine again, but in practice will have tiny off diagonal elements (e.g. ~1e-15) due to finite floating point precision. This scenario is covered in one of the test cases here.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
maybe more of a "performance optimization" than new feature as end users would typically not need to interact with this at all.

# References
split off from #4303 as it is not strictly related to permutations. The caching used only for the `is_diagonal` property here also is used for a couple of other properties in that PR.

# How has this been tested?

- [x] existing tests pass
- [x] new tests were added

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/transforms/transforms.py [hunk 1]
```diff
@@ -469,6 +491,8 @@ def shear(self, shear):
             self.linear_matrix, upper_triangular=self._upper_triangular
         )
         self._linear_matrix = compose_linear_matrix(rotate, scale, shear)
+        # previously cached properties may have changed
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Optional: consider removing this implementation comment, because I think the call to `_clean_cache` implies this pretty clearly and because if you need it here you probably need it everywhere you call `_clean_cache` which would be unfortunate.

Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Optional: consider removing this implementation comment, because I think the call to `_clean_cache` implies this pretty clearly and because if you need it here you probably need it everywhere you call `_clean_cache` which would be unfortunate.

File: napari/utils/transforms/transforms.py [hunk 2]
```diff
@@ -597,6 +628,14 @@ def expand_dims(self, axes: Sequence[int]) -> 'Affine':
             name=self.name,
         )
 
+    @cached_property
+    def is_diagonal(self):
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: How would you feel about adding an underscore to `is_diagonal`, so that it isn't explicitly part of the public API? At the moment, this property itself is being used as an optimization, and I'm a little unsure if it's something we'll want to support in the long term future if we end up re-designing transforms as in, so I'd lean towards being conservative and not expanding the already large public API here.

I don't feel too strongly about this, as I guess lots of transforms could just return `False` as a default implementation.

Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: How would you feel about adding an underscore to `is_diagonal`, so that it isn't explicitly part of the public API? At the moment, this property itself is being used as an optimization, and I'm a little unsure if it's something we'll want to support in the long term future if we end up re-designing transforms as in, so I'd lean towards being conservative and not expanding the already large public API here.

I don't feel too strongly about this, as I guess lots of transforms could just return `False` as a default implementation.

File: napari/utils/transforms/transforms.py [hunk 3]
```diff
@@ -119,6 +121,12 @@ def inverse(self) -> 'TransformChain':
         """Return the inverse transform chain."""
         return TransformChain([tf.inverse for tf in self[::-1]])
 
+    @property
+    def is_diagonal(self):
+        if all(getattr(tf, 'is_diagonal', False) for tf in self):
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Instead of using a default of `False` for `getattr`, what do you think about defining `is_diagonal` in the base `Transform` class and having it return `False`?

Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Instead of using a default of `False` for `getattr`, what do you think about defining `is_diagonal` in the base `Transform` class and having it return `False`?

File: napari/utils/transforms/transform_utils.py [hunk 4]
```diff
@@ -466,3 +466,27 @@ def is_matrix_triangular(matrix):
     return is_matrix_upper_triangular(matrix) or is_matrix_lower_triangular(
         matrix
     )
+
+
+def is_diagonal(matrix, tol=1e-8):
+    """Determine whether affine is a diagonal matrix.
```
Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
    """Determine whether a matrix is diagonal up to some tolerance.
```

No need to mention affine here, especially as we describe affine transforms matrices with non-zero translation to be diagonal.

Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
    """Determine whether a square matrix is diagonal up to some tolerance.
```

No need to mention affine here, especially as we describe affine transforms matrices with non-zero translation to be diagonal.

File: napari/utils/transforms/transform_utils.py [hunk 5]
```diff
@@ -466,3 +466,27 @@ def is_matrix_triangular(matrix):
...
+    ----------
+    matrix : 2-D array
+        The matrix to test.
+    tol : float, optional
+        Consider any entries with magnitude < `tol` as 0.
+
+    Returns
+    -------
+    is_diag : bool
+        Boolean indicating whether affine is diagonal.
```
Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
        True if matrix is diagonal, False otherwise.
```

Optional: I prefer being explicit about True/False assignments instead of using whether/if, but not required. You probably should change `affine` to `matrix` though.

Comment (andy-sweet) [inline napari/utils/transforms/transform_utils.py]:
```suggestion
        True if matrix is diagonal, False otherwise.
```

Optional: I prefer being explicit about True/False assignments instead of using whether/if, but not required. You probably should change `affine` to `matrix` though.

File: napari/utils/transforms/_tests/test_transforms.py [hunk 6]
```diff
@@ -18,6 +18,26 @@ def test_scale_translate(Transform):
...
+    transform.rotate = 5.0
+    assert not transform.is_diagonal
+    # Rotation back to 0.0 will result in tiny non-zero off-diagonal values.
+    # is_diagonal assumes values below 1e-8 are equivalent to 0.
+    transform.rotate = 0.0
+    assert transform.is_diagonal
+
+
+def test_diagonal_scale_setter():
+    # diagonal matrices are also considered a permutation
```
Comment (andy-sweet) [inline napari/utils/transforms/_tests/test_transforms.py]: Is this comment left over from the larger PR? Permutations don't seem related to this test as is.

Comment (andy-sweet) [inline napari/utils/transforms/_tests/test_transforms.py]: Is this comment left over from the larger PR? Permutations don't seem related to this test as is.

File: napari/utils/transforms/transforms.py [hunk 7]
```diff
@@ -195,6 +203,7 @@ def __init__(self, scale=(1.0,), translate=(0.0,), *, name=None):
 
         self.scale = np.array(scale)
         self.translate = np.array(translate)
+        self.is_diagonal = True
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: I guess it's OK to have different ways to implement `is_diagonal` (i.e. here a plain attribute, vs. a `@cached_property` in `Affine`), but it's a little confusing.

Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: I guess it's OK to have different ways to implement `is_diagonal` (i.e. here a plain attribute, vs. a `@cached_property` in `Affine`), but it's a little confusing.

File: napari/utils/transforms/transforms.py [hunk 8]
```diff
@@ -597,6 +628,14 @@ def expand_dims(self, axes: Sequence[int]) -> 'Affine':
             name=self.name,
         )
 
+    @cached_property
+    def is_diagonal(self):
+        return is_diagonal(self.linear_matrix, tol=1e-8)
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Would be good to document that the diagonal check does not include the translation component of the whole affine matrix.

Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: Would be good to document that the diagonal check does not include the translation component of the whole affine matrix.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Mostly looks good!

Overall, I lean towards defining `Transform._is_diagonal` as a non-public property with a default implementation that returns `False` and that can be overridden by subclasses. For me that makes things safer and more clear.

That would also allow us to document this property in one place to define exactly what it means. But that might also require a rename since in general a `Transform` may not even have a matrix or sub-matrix to check (maybe something like `_mixes_dimensions` but better?). So, I'd be open to other ideas here too.

File: napari/utils/transforms/transforms.py (hunk 2 above)
Comment (grlee77) [reply napari/utils/transforms/transforms.py]: Yeah, I am fine with it being private

File: napari/utils/transforms/_tests/test_transforms.py (hunk 6 above)
Comment (grlee77) [reply napari/utils/transforms/_tests/test_transforms.py]: yes, I will remove it

**Comment by grlee77** [comment]
Comments have been addressed aside from deciding on a potential different name like `_mixes_dimensions` or similar.

File: napari/utils/transforms/transforms.py [hunk 9]
```diff
@@ -272,6 +293,11 @@ def expand_dims(self, axes: Sequence[int]) -> 'ScaleTranslate':
         translate[not_axes] = self.translate
         return ScaleTranslate(scale, translate, name=self.name)
 
+    @cached_property
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
    @property
```

Nit: this always returns true, so just a plain property seems fine.

File: napari/utils/transforms/transforms.py [hunk 10]
```diff
@@ -93,6 +95,19 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
             trans._('Cannot subset arbitrary transforms.', deferred=True)
         )
 
+    @cached_property
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
    @property
```

Nit: this always returns false, so just a plain property seems fine.

**Review by andy-sweet — APPROVED** [review]
Looks good to me now.

My only nit is to use property instead of cached_property when that's enough. Unless there's something I don't understand about overriding those?

As this `_is_diagonal` now, it's name is not as important and I think the docstring explains it well enough, so need to think of other names right now.

You should get a core-dev to look at this who can approve/merge this.

**Comment by andy-sweet** [comment]
Some of my comments are being rendered strangely / duplicated (at least on my end) - sorry about that...

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
I'll merge this in ~24h if no one weighs in before then. Thanks @grlee77 and @andy-sweet!
